Make sure to evaluate annotation strings in the context of where the prototype was created

aorenste opened the pull request:
This was incorrectly evaluating the annotation in the context of infer_schema - make sure to evaluate annotation strings in the context of where the prototype was created instead.

Fixes #145481 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145667

aorenste committed:
Make sure to evaluate annotation strings in the context of where the prototype was created

[ghstack-poisoned]

aorenste committed:
Update on "Make sure to evaluate annotation strings in the context of where the prototype was created"




[ghstack-poisoned]

aorenste committed:
Update on "Make sure to evaluate annotation strings in the context of where the prototype was created"

This was incorrectly evaluating the annotation in the context of infer_schema - make sure to evaluate annotation strings in the context of where the prototype was created instead.

Fixes #145481 




[ghstack-poisoned]

zou3519 commented on torch/_library/infer_schema.py:
```diff
@@ -53,6 +53,13 @@ def infer_schema(
     """
     UNKNOWN_MUTATES = "unknown"
+    pf_globals = prototype_function.__globals__
+    # To do this properly we really need `locals()` from the scope of the
+    # prototype_function. Unfortunately there doesn't seem to be any way to get
+    # that consistently.
+    pf_locals = None
```
prototype_function is always a global function (i.e. not a nested function). So the locals() for it is actually the same as globals(). Does that sound right?

zou3519 replied:
Actually I take that back, it's not always a global function.

zou3519 commented on torch/_library/infer_schema.py:
```diff
@@ -53,6 +53,13 @@ def infer_schema(
+    # prototype_function. Unfortunately there doesn't seem to be any way to get
+    # that consistently.
+    pf_locals = None
+    # TODO: py3.10+ pass the `globals` parameter and we no longer need to deal
+    # with stringified annotations.
```
What do you mean?

aorenste replied:
As of py3.10+ inspect.signature() accepts a `globals` (and `locals`) parameter which if specified will do the eval() within the `inspect.signature` itself so it will never return stringified types.

zou3519 commented on torch/_library/infer_schema.py:
```diff
@@ -62,7 +69,7 @@ def error_fn(what):
     def convert_type_string(annotation_type: str):
         try:
-            return eval(annotation_type)
+            return eval(annotation_type, pf_globals, pf_locals)
```
Is it possible to add a test?

aorenste committed:
Update on "Make sure to evaluate annotation strings in the context of where the prototype was created"

This was incorrectly evaluating the annotation in the context of infer_schema - make sure to evaluate annotation strings in the context of where the prototype was created instead.

Fixes #145481 




[ghstack-poisoned]

zou3519 reviewed (approved):
thank you

aorenste commented:
@pytorchbot merge

aorenste commented:
@pytorchbot merge -i

aorenste commented:
@pytorchbot merge -i
